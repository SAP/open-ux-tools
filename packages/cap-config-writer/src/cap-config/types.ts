export type CdsUi5PluginInfo = {
    /**
     * Convienience property. The CDS UI5 plugin is considered enabled if `hasCdsUi5Plugin`, `hasMinCdsVersion`, `isWorkspaceEnabled` are all true.
     */
    isCdsUi5PluginEnabled: boolean;
    /**
     * True if the CDS version satisfies the minimum supported CDS version
     */
    hasMinCdsVersion: boolean;
    /**
     * True if NPM workspaces are enabled at the root of a CAP project
     */
    isWorkspaceEnabled: boolean;
    /**
     * True if the CDS ui5 plugin is specified as a dependency
     */
    hasCdsUi5Plugin: boolean;
};
