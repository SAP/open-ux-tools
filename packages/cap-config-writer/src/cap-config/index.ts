import { join } from 'path';
import { create as createStorage } from 'mem-fs';
import { create } from 'mem-fs-editor';
import type { Editor } from 'mem-fs-editor';
import type { Package } from '@sap-ux/project-access';
import {
    addCdsPluginUi5,
    enableWorkspaces,
    ensureMinCdsVersion,
    getWorkspaceInfo,
    hasCdsPluginUi5,
    hasMinCdsVersion
} from './package-json';

/**
 * Enable workspace and cds-plugin-ui5 for given CAP project.
 *
 * @param basePath - root path of the CAP project, where package.json is located
 * @param [fs] - optional: the memfs editor instance
 * @returns Promise<Editor> - memfs editor instance with updated files
 */
export async function enableCdsUi5Plugin(basePath: string, fs?: Editor): Promise<Editor> {
    if (!fs) {
        fs = create(createStorage());
    }
    const packageJsonPath = join(basePath, 'package.json');
    const packageJson = (fs.readJSON(packageJsonPath) ?? {}) as Package;

    ensureMinCdsVersion(packageJson);
    await enableWorkspaces(basePath, packageJson);
    addCdsPluginUi5(packageJson);

    fs.writeJSON(packageJsonPath, packageJson);
    return fs;
}

/**
 * Check if cds-plugin-ui5 is enabled on a CAP project. Checks also all prerequisites, like minimum @sap/cds version.
 *
 * @param basePath - root path of the CAP project, where package.json is located
 * @param [fs] - optional: the memfs editor instance
 * @returns true: cds-plugin-ui5 and all prerequisites are fulfilled; false: cds-plugin-ui5 is not enabled or not all prerequisites are fulfilled
 */
export async function checkCdsUi5PluginEnabled(basePath: string, fs?: Editor): Promise<boolean> {
    if (!fs) {
        fs = create(createStorage());
    }
    const packageJsonPath = join(basePath, 'package.json');
    if (!fs.exists(packageJsonPath)) {
        return false;
    }
    const packageJson = fs.readJSON(packageJsonPath) as Package;
    const { workspaceEnabled } = await getWorkspaceInfo(basePath, packageJson);
    return hasMinCdsVersion(packageJson) && workspaceEnabled && hasCdsPluginUi5(packageJson);
}
