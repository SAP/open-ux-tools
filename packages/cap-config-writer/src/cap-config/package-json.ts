import { coerce, gte, satisfies } from 'semver';
import { getCapCustomPaths } from '@sap-ux/project-access';
import type { Package } from '@sap-ux/project-access';

const minCdsVersion = '6.8.2';
const minCdsPluginUi5Version = '0.1.4';

/**
 * Ensure a minimum version of @sap/cds in dependencies.
 *
 * @param packageJson - the parsed package.json
 */
export function ensureMinCdsVersion(packageJson: Package): void {
    if (!hasMinCdsVersion(packageJson)) {
        packageJson.dependencies ??= {};
        packageJson.dependencies['@sap/cds'] = `^${minCdsVersion}`;
    }
}

/**
 * Enable workspaces for app/* folders.
 *
 * @param basePath - root path of the CAP project, where package.json is located
 * @param packageJson - the parsed package.json
 */
export async function enableWorkspaces(basePath: string, packageJson: Package): Promise<void> {
    const { appWorkspace, workspaceEnabled } = await getWorkspaceInfo(basePath, packageJson);
    if (workspaceEnabled) {
        return;
    }
    let workspacePackages = getWorkspacePackages(packageJson);
    if (!workspacePackages) {
        packageJson.workspaces ??= [];
        if (Array.isArray(packageJson.workspaces)) {
            workspacePackages = packageJson.workspaces;
        } else {
            packageJson.workspaces.packages = [];
            workspacePackages = packageJson.workspaces.packages;
        }
    }
    workspacePackages.push(appWorkspace);
}

/**
 * Add devDependency to cds-plugin-ui5.
 *
 * @param packageJson - the parsed package.json
 */
export function addCdsPluginUi5(packageJson: Package): void {
    if (!hasCdsPluginUi5(packageJson)) {
        packageJson.devDependencies ??= {};
        packageJson.devDependencies['cds-plugin-ui5'] = `^${minCdsPluginUi5Version}`;
    }
}

/**
 * Check if package.json has dependency to the minimum min version of @sap/cds,
 * that is required to enable cds-plugin-ui.
 *
 * @param packageJson - the parsed package.json
 * @returns - true: min cds version is present; false: cds version needs update
 */
export function hasMinCdsVersion(packageJson: Package): boolean {
    return gte(coerce(packageJson.dependencies?.['@sap/cds']) ?? '0.0.0', minCdsVersion);
}

/**
 * Check if package.json has version or version range that satisfies the minimum version of @sap/cds.
 *
 * @param packageJson  - the parsed package.json
 * @returns - true: cds version satisfies the min cds version; false: cds version does not satisfy min cds version
 */
export function satisfiesMinCdsVersion(packageJson: Package): boolean {
    return hasMinCdsVersion(packageJson) || satisfies(minCdsVersion, packageJson.dependencies?.['@sap/cds'] ?? '0.0.0');
}

/**
 * Get information about the workspaces in the CAP project.
 *
 * @param basePath - root path of the CAP project, where package.json is located
 * @param packageJson - the parsed package.json
 * @returns - appWorkspace containing the path to the appWorkspace including wildcard; workspaceEnabled: boolean that states whether workspace for apps are enabled
 */
export async function getWorkspaceInfo(
    basePath: string,
    packageJson: Package
): Promise<{ appWorkspace: string; workspaceEnabled: boolean }> {
    const capPaths = await getCapCustomPaths(basePath);
    const appWorkspace = capPaths.app.endsWith('/') ? `${capPaths.app}*` : `${capPaths.app}/*`;
    const workspacePackages = getWorkspacePackages(packageJson) ?? [];
    const workspaceEnabled = workspacePackages.includes(appWorkspace);
    return { appWorkspace, workspaceEnabled };
}

/**
 * Return the reference to the array of workspace packages or undefined if not defined.
 * The workspace packages can either be defined directly as workspaces in package.json
 * or in workspaces.packages, e.g. in yarn workspaces.
 *
 * @param packageJson - the parsed package.json
 * @returns ref to the packages in workspaces or undefined
 */
function getWorkspacePackages(packageJson: Package): string[] | undefined {
    let workspacePackages: string[] | undefined;
    if (Array.isArray(packageJson.workspaces)) {
        workspacePackages = packageJson.workspaces;
    } else if (Array.isArray(packageJson.workspaces?.packages)) {
        workspacePackages = packageJson.workspaces?.packages;
    }
    return workspacePackages;
}

/**
 * Check if devDependency to cds-plugin-ui5 is present in package.json.
 *
 * @param packageJson - the parsed package.json
 * @returns true: devDependency to cds-plugin-ui5 exists; false: devDependency to cds-plugin-ui5 does not exist
 */
export function hasCdsPluginUi5(packageJson: Package): boolean {
    return !!packageJson.devDependencies?.['cds-plugin-ui5'];
}
