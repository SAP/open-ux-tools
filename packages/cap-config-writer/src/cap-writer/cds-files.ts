import { toReferenceUri } from '@sap-ux/project-access';
import type { Editor } from 'mem-fs-editor';
import { t } from '../i18n';
import { join } from 'path';
import { type CapService } from '../types/capTypes';
import { getAnnotationPath } from '@sap-ux/fiori-generator-shared';
import type { Logger } from '@sap-ux/logger';
import { toPosixPath } from './helpers';

/**
 * Updates either the index.cds or services.cds file with the provided annotation configuration.
 * If neither file exists, a new services.cds file is created with the annotation configuration.
 *
 * @param {Editor} fs - The file system editor.
 * @param {string} annotationConfig - The annotation configuration to append to the file.
 * @param {CapService} capService - The CAP service configuration.
 */
async function updateCdsIndexOrServiceFile(fs: Editor, annotationConfig: string, capService: CapService) {
    const indexFile = join(capService.projectPath, capService.appPath ?? '', 'index.cds');
    const serviceFile = join(capService.projectPath, capService.appPath ?? '', 'services.cds');

    if (fs.exists(indexFile)) {
        fs.append(indexFile, annotationConfig);
    } else if (fs.exists(serviceFile)) {
        fs.append(serviceFile, annotationConfig);
    } else {
        fs.write(serviceFile, annotationConfig);
    }
}

/**
 * Generates the annotation configuration string for a given project name.
 * This configuration string is used to import annotations from the project's 'annotations' directory.
 *
 * @param {string} projectName - The name of the project.
 * @returns {string} The annotation configuration string.
 */
function generateAnnotationConfig(projectName: string): string {
    return `\nusing from './${toPosixPath(join(projectName, 'annotations'))}';`;
}

/**
 * Writes the annotation CDS file with a given service name and service CDS path.
 *
 * @param {Editor} fs - The file system editor.
 * @param {string} annotationCdsPath - The path to the annotation CDS file.
 * @param {string} serviceName - The name of the service.
 * @param {string} serviceCdsPath - The path to the service CDS file.
 */
async function writeAnnotationCdsFile(
    fs: Editor,
    annotationCdsPath: string,
    serviceName: string,
    serviceCdsPath: string
) {
    fs.write(annotationCdsPath, `using ${serviceName} as service from '${serviceCdsPath}';`);
}

/**
 * Updates CAP CDS files by adding annotation references.
 *
 * @param {Editor} fs - The file system editor.
 * @param {CapService} capService - The CAP service configuration.
 * @param {string} projectName - The name of the project.
 * @param {Logger} [logger] - The logger instance for logging messages.
 * @returns {Promise<void>} A Promise that resolves when the operation is complete.
 */
export async function updateCdsFilesWithAnnotations(
    fs: Editor,
    capService: CapService,
    projectName: string,
    logger?: Logger
): Promise<void> {
    const annotationPath = getAnnotationPath(projectName, capService?.appPath);
    const serviceCdsPath = await toReferenceUri(
        capService.projectPath,
        annotationPath,
        capService.serviceCdsPath ?? ''
    );
    const annotationConfig = generateAnnotationConfig(projectName);
    const annotationCdsPath = join(capService.projectPath, capService.appPath ?? '', projectName, 'annotations.cds');
    logger?.info(t('info.capServiceName', { capServiceName: capService.serviceName }));
    logger?.info(
        t('info.cdsUpdateInfo', {
            projectPath: {
                root: capService.projectPath
            },
            annotationPath: annotationPath,
            capService: capService
        })
    );
    await writeAnnotationCdsFile(fs, annotationCdsPath, capService.serviceName, serviceCdsPath);
    await updateCdsIndexOrServiceFile(fs, annotationConfig, capService);
}
