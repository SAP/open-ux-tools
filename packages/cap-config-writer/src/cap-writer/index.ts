export { updateCdsFilesWithAnnotations } from './cds-files';
export { updateAppPackageJsonCAP, updateRootPackageJsonCAP } from './package-json';
export { updateCAPManifestJson } from './manifest-json';
export { generateCapLaunchText, getCDSTask } from './helpers';
export { updateTsConfigCap } from './ts-config';
export { updatePomXml } from './pom-xml';
