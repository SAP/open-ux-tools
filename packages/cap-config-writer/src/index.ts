export { checkCdsUi5PluginEnabled, enableCdsUi5Plugin, hasMinCdsVersion } from './cap-config';
