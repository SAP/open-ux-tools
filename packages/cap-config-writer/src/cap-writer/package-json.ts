import type { Package } from '@sap-ux/project-access';
import { getCapCustomPaths, checkCdsUi5PluginEnabled } from '@sap-ux/project-access';
import type { Editor } from 'mem-fs-editor';
import { dirname, join, normalize, posix } from 'path';
import type { CapServiceCdsInfo } from '../cap-config/types';
import { enableCdsUi5Plugin } from '../cap-config';
import type { Logger } from '@sap-ux/logger';

/**
 * Retrieves the CDS watch script for the CAP app.
 *
 * @param {string} projectName - The project's name, which is the module name.
 * @param {string} appId - The application's ID, including its namespace and the module name.
    If appId is provided, it will be used to open the application instead of the project name. This option is available for use with npm workspaces.
 * @returns {{ [x: string]: string }} The CDS watch script for the CAP app.
 */
export function getCDSWatchScript(projectName: string, appId?: string): { [x: string]: string } {
    const DisableCacheParam = 'sap-ui-xx-viewCache=false';
    // projects by default are served base on the folder name in the app/ folder
    const project = appId ?? projectName + '/webapp';
    const watchScript = {
        [`watch-${projectName}`]: `cds watch --open ${project}/index.html?${DisableCacheParam}${
            appId ? ' --livereload false' : ''
        }`
    };
    return watchScript;
}

/**
 * Updates the scripts in the package json file with the provided scripts object.
 *
 * @param {Editor} fs - The file system editor.
 * @param {string} packageJsonPath - The path to the package.json file.
 * @param {Record<string, string>} scripts - The scripts to be added or updated in the package.json file.
 * @returns {void}
 */
function updatePackageJsonWithScripts(fs: Editor, packageJsonPath: string, scripts: Record<string, string>): void {
    fs.extendJSON(packageJsonPath, { scripts });
}

/**
 * Updates the scripts in the package json file for a CAP project.
 *
 * @param {Editor} fs - The file system editor.
 * @param {string} packageJsonPath - The path to the package.json file.
 * @param {string} projectName - The project's name, which is the module name.
 * @param {string} appId - The application's ID, including its namespace and the module name.
 * @param {boolean} [enableNPMWorkspaces] - Whether to enable npm workspaces.
 * @returns {Promise<void>} A Promise that resolves once the scripts are updated.
 */
async function updateScripts(
    fs: Editor,
    packageJsonPath: string,
    projectName: string,
    appId: string,
    enableNPMWorkspaces?: boolean
): Promise<void> {
    const hasNPMworkspaces = await checkCdsUi5PluginEnabled(dirname(packageJsonPath), fs);
    let cdsScript;
    // eslint-disable-next-line @typescript-eslint/prefer-nullish-coalescing
    if (enableNPMWorkspaces || hasNPMworkspaces) {
        // If the project uses npm workspaces (and specifically cds-plugin-ui5 ) then the project is served using the appId
        cdsScript = getCDSWatchScript(projectName, appId);
    } else {
        cdsScript = getCDSWatchScript(projectName);
    }
    updatePackageJsonWithScripts(fs, packageJsonPath, cdsScript);
}

/**
 * Updates the root package.json file of CAP projects with the following changes:
 * 1) Adds the app name to the sapux array in the root package.json if sapux is enabled.
 * 2) Adds the cds watch script to the root package.json if applicable.
 *
 * @param {Editor} fs - The file system editor.
 * @param {string} projectName - The project's name, which is the module name.
 * @param {boolean} sapux - Whether to add the app name to the sapux array.
 * @param {CapServiceCdsInfo} capService - The CAP service instance.
 * @param {string} appId - The application's ID, including its namespace and the module name.
 * @param {Logger} [log] - The logger instance for logging warnings.
 * @param {boolean} [enableNPMWorkspaces] - Whether to enable npm workspaces.
 * @returns {Promise<void>} A Promise that resolves once the root package.json is updated.
 */
export async function updateRootPackageJson(
    fs: Editor,
    projectName: string,
    sapux: boolean,
    capService: CapServiceCdsInfo,
    appId: string,
    log?: Logger,
    enableNPMWorkspaces?: boolean
): Promise<void> {
    const packageJsonPath: string = join(capService.projectPath, 'package.json');
    const packageJson = (fs.readJSON(packageJsonPath) ?? {}) as Package;
    const capNodeType = 'Node.js';

    if (enableNPMWorkspaces && packageJson) {
        await enableCdsUi5Plugin(capService.projectPath, fs);
    }
    if (capService?.capType === capNodeType) {
        await updateScripts(fs, packageJsonPath, projectName, appId, enableNPMWorkspaces);
    }
    if (sapux) {
        const dirPath = join(capService.appPath ?? (await getCapCustomPaths(capService.projectPath)).app, projectName);
        // Converts a directory path to a POSIX-style path.
        const capProjectPath = normalize(dirPath).split(/[\\/]/g).join(posix.sep);
        const sapuxExt = Array.isArray(packageJson?.sapux) ? [...packageJson.sapux, capProjectPath] : [capProjectPath];
        fs.extendJSON(packageJsonPath, { sapux: sapuxExt });
    }
}

/**
 * Updates the package.json file of a CAP project app by removing the sapux property
 * and start scripts, as well as the 'int-test' script.
 *
 * @param {Editor} fs The file system editor.
 * @param {string} appRoot The root directory of the application.
 */
export function updateAppPackageJson(fs: Editor, appRoot: string): void {
    const packageJsonPath: string = join(appRoot, 'package.json');
    const packageJson = (fs.readJSON(packageJsonPath) ?? {}) as Package;

    delete packageJson.sapux;
    if (packageJson?.scripts) {
        delete packageJson.scripts['int-test'];
    }
    for (const script in packageJson.scripts) {
        if (script.startsWith('start')) {
            delete packageJson.scripts[script];
        }
    }
    fs.writeJSON(packageJsonPath, packageJson);
}
