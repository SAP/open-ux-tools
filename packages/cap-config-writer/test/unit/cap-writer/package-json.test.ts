import type { CapRuntime } from '@sap-ux/odata-service-inquirer';
import { satisfiesMinCdsVersion } from '../../../src/cap-config/package-json';
import memFs from 'mem-fs';
import { ToolsLogger } from '@sap-ux/logger';
import editor, { type Editor } from 'mem-fs-editor';
import { join } from 'path';
import { minCdsVersion } from '../../../src/cap-config';
import { updateRootPackageJson, updateAppPackageJson } from '../../../src/cap-writer/package-json';
import type { Package } from '@sap-ux/project-access';
import type { CapServiceCdsInfo } from '../../../src/index';

jest.mock('../../../src/cap-config/package-json', () => ({
    ...jest.requireActual('../../../src/cap-config/package-json'),
    satisfiesMinCdsVersion: jest.fn()
}));

jest.mock('@sap-ux/project-access', () => ({
    ...jest.requireActual('@sap-ux/project-access'),
    getCdsVersionInfo: jest.fn()
}));

describe('Writing/package json files', () => {
    let fs: Editor;
    const testInputPath = join(__dirname, 'test-inputs');
    const logger = new ToolsLogger();
    const testProjectNameNoSapUx = 'test-cap-package-no-sapux';
    const testProjectNameWithSapUx = 'test-cap-package-sapux';
    let capService: CapServiceCdsInfo;
    const capNodeType: CapRuntime = 'Node.js';

    // beforeEach function to reset fs before each test
    beforeEach(() => {
        const store = memFs.create();
        // Create a new instance of the Editor class before each test
        fs = editor.create(store);
    });

    beforeEach(() => {
        capService = {
            projectPath: join(testInputPath, testProjectNameNoSapUx),
            serviceName: 'AdminService',
            serviceCdsPath: 'srv/admin-service',
            appPath: 'app',
            capType: capNodeType,
            cdsUi5PluginInfo: {
                isCdsUi5PluginEnabled: false,
                hasMinCdsVersion: true,
                isWorkspaceEnabled: false,
                hasCdsUi5Plugin: false
            }
        };
    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    test('should update package scripts for cap projects with sap ux enabled', async () => {
        const packageJsonPath = join(testInputPath, testProjectNameWithSapUx, 'package.json');
        const isSapUxEnabled = true;
        capService.projectPath = join(testInputPath, testProjectNameWithSapUx);
        (satisfiesMinCdsVersion as jest.Mock).mockReturnValue(true);
        await updateRootPackageJson(fs, testProjectNameNoSapUx, isSapUxEnabled, capService, 'test.app.project');
        const packageJson = (fs.readJSON(packageJsonPath) ?? {}) as Package;
        const scripts = packageJson.scripts;
        expect(scripts).toEqual({
            'watch-test-cap-package-no-sapux':
                'cds watch --open test-cap-package-no-sapux/webapp/index.html?sap-ui-xx-viewCache=false'
        });
    });

    test('should log warning if no minimum cds version is being satisfied', async () => {
        const testProjectMinCds = 'test-cap-package-no-min-cds-version';
        const isSapUxEnabled = false;
        const logger = new ToolsLogger();
        const loggerMock = jest.fn();
        logger.warn = loggerMock;
        (satisfiesMinCdsVersion as jest.Mock).mockReturnValue(false);
        const capServiceWithNoMinCds = capService;
        capServiceWithNoMinCds.cdsUi5PluginInfo.hasMinCdsVersion = false;
        await updateRootPackageJson(
            fs,
            testProjectMinCds,
            isSapUxEnabled,
            capServiceWithNoMinCds,
            'test.app.project',
            logger
        );
        expect(logger.warn).toHaveBeenCalledTimes(1);
        expect(logger.warn).toHaveBeenCalledWith(
            expect.stringContaining(`minimum cds-dk ${minCdsVersion} version is required to add cds watch scripts`)
        );
    });

    test('should enable CdsUi5Plugin when workspace is enabled', async () => {
        const isSapUxEnabled = true;
        const isNpmWorkspacesEnabled = true;
        const packageJsonPath = join(testInputPath, testProjectNameNoSapUx, 'package.json');
        await updateRootPackageJson(
            fs,
            testProjectNameWithSapUx,
            isSapUxEnabled,
            capService,
            'test.app.project',
            logger,
            isNpmWorkspacesEnabled
        );
        const packageJson = (fs.readJSON(packageJsonPath) ?? {}) as Package;
        const devDependencies = packageJson.devDependencies;
        expect(devDependencies).toEqual({
            'cds-plugin-ui5': '^0.6.13'
        });
    });

    test('should remove int-test script and start scripts, and also keep other scripts', async () => {
        const appRoot = join(__dirname, 'test-inputs/test-cap-package-no-sapux');
        const packageJsonPath = join(capService.projectPath, 'package.json');
        updateAppPackageJson(fs, appRoot);
        const packageJson = (fs.readJSON(packageJsonPath) ?? {}) as Package;
        const scripts = packageJson.scripts;
        expect(scripts).toEqual({
            'test-script': 'Run some scripts here'
        });
        expect(scripts).not.toHaveProperty('int-test');
        expect(scripts).not.toHaveProperty('start');
    });

    test('should remove int-test script and start scripts, and also keep other scripts', async () => {
        const appRoot = join(__dirname, 'test-inputs/test-cap-package-no-sapux');
        const packageJsonPath = join(capService.projectPath, 'package.json');
        updateAppPackageJson(fs, appRoot);
        const packageJson = (fs.readJSON(packageJsonPath) ?? {}) as Package;
        const scripts = packageJson.scripts;
        expect(scripts).toEqual({
            'test-script': 'Run some scripts here'
        });
        expect(scripts).not.toHaveProperty('int-test');
        expect(scripts).not.toHaveProperty('start');
    });
});
