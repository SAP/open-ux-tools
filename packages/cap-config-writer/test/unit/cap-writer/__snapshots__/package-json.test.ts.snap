// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Writing/package json files should enable CdsUi5Plugin when workspace is enabled 1`] = `
Object {
  "": Object {
    "contents": "{
  \\"name\\": \\"test_cap\\",
  \\"version\\": \\"0\\",
  \\"description\\": \\"Test file\\",
  \\"dependencies\\": {
    \\"@sap/cds\\": \\"6.8.2\\"
  },
  \\"sapux\\": [
    \\"app/test-cap-package-no-sapux\\"
  ],
  \\"scripts\\": {
    \\"watch-test-cap-package-no-sapux\\": \\"cds watch --open test-cap-package-no-sapux/webapp/index.html?sap-ui-xx-viewCache=false\\"
  }
}
",
    "state": "modified",
  },
  "../../test-cap-package-no-sapux/package.json": Object {
    "contents": "{
  \\"name\\": \\"test_cap\\",
  \\"version\\": \\"0\\",
  \\"description\\": \\"Test file\\",
  \\"scripts\\": {
    \\"test-script\\": \\"Run some scripts here\\",
    \\"int-test\\": \\"test command\\",
    \\"start\\": \\"start command\\"
  },
  \\"dependencies\\": {
    \\"@sap/cds\\": \\"^6.8.2\\"
  },
  \\"workspaces\\": [
    \\"app/*\\"
  ],
  \\"devDependencies\\": {
    \\"cds-plugin-ui5\\": \\"^0.6.13\\"
  },
  \\"sapux\\": [
    \\"app/test-cap-package-sapux\\"
  ]
}
",
    "state": "modified",
  },
}
`;

exports[`Writing/package json files should log warning if no minimum cds version is being satisfied 1`] = `
Object {
  "../../test-cap-package-sapux/package.json": Object {
    "contents": "{
  \\"name\\": \\"test_cap\\",
  \\"version\\": \\"0\\",
  \\"description\\": \\"Test file\\",
  \\"dependencies\\": {
    \\"@sap/cds\\": \\"6.8.2\\"
  },
  \\"sapux\\": [
    \\"app/test-cap-package-no-sapux\\"
  ],
  \\"scripts\\": {
    \\"watch-test-cap-package-no-sapux\\": \\"cds watch --open test-cap-package-no-sapux/webapp/index.html?sap-ui-xx-viewCache=false\\"
  }
}
",
    "state": "modified",
  },
}
`;

exports[`Writing/package json files should remove int-test script and keep other scripts 1`] = `
Object {
  "": Object {
    "contents": "{
  \\"name\\": \\"test_cap\\",
  \\"version\\": \\"0\\",
  \\"description\\": \\"Test file\\",
  \\"scripts\\": {
    \\"test-script\\": \\"Run some scripts here\\"
  },
  \\"dependencies\\": {
    \\"@sap/cds\\": \\"^6.8.2\\"
  },
  \\"workspaces\\": [
    \\"app/*\\"
  ],
  \\"devDependencies\\": {
    \\"cds-plugin-ui5\\": \\"^0.6.13\\"
  }
}
",
    "state": "modified",
  },
  "../../../../../../app/package.json": Object {
    "contents": "{}
",
    "state": "modified",
  },
  "../../test-cap-package-sapux/package.json": Object {
    "contents": "{
  \\"name\\": \\"test_cap\\",
  \\"version\\": \\"0\\",
  \\"description\\": \\"Test file\\",
  \\"dependencies\\": {
    \\"@sap/cds\\": \\"6.8.2\\"
  },
  \\"sapux\\": [
    \\"app/test-cap-package-no-sapux\\"
  ],
  \\"scripts\\": {
    \\"watch-test-cap-package-no-sapux\\": \\"cds watch --open test-cap-package-no-sapux/webapp/index.html?sap-ui-xx-viewCache=false\\"
  }
}
",
    "state": "modified",
  },
}
`;

exports[`Writing/package json files should remove sapux property from package json and start scripts 1`] = `
Object {
  "../../app/package.json": Object {
    "contents": "{}
",
    "state": "modified",
  },
  "../../test/unit/cap-writer/test-inputs/test-cap-package-no-sapux/package.json": Object {
    "contents": "{
  \\"name\\": \\"test_cap\\",
  \\"version\\": \\"0\\",
  \\"description\\": \\"Test file\\",
  \\"scripts\\": {
    \\"test-script\\": \\"Run some scripts here\\",
    \\"int-test\\": \\"test command\\",
    \\"start\\": \\"start command\\"
  },
  \\"dependencies\\": {
    \\"@sap/cds\\": \\"^6.8.2\\"
  },
  \\"workspaces\\": [
    \\"app/*\\"
  ],
  \\"devDependencies\\": {
    \\"cds-plugin-ui5\\": \\"^0.6.13\\"
  },
  \\"sapux\\": [
    \\"app/test-cap-package-sapux\\"
  ]
}
",
    "state": "modified",
  },
  "../../test/unit/cap-writer/test-inputs/test-cap-package-sapux/package.json": Object {
    "contents": "{
  \\"name\\": \\"test_cap\\",
  \\"version\\": \\"0\\",
  \\"description\\": \\"Test file\\",
  \\"dependencies\\": {
    \\"@sap/cds\\": \\"6.8.2\\"
  },
  \\"sapux\\": [
    \\"app/test-cap-package-no-sapux\\"
  ],
  \\"scripts\\": {
    \\"watch-test-cap-package-no-sapux\\": \\"cds watch --open test-cap-package-no-sapux/webapp/index.html?sap-ui-xx-viewCache=false\\"
  }
}
",
    "state": "modified",
  },
}
`;

exports[`Writing/package json files should update package for cap projects with sap ux disabled 1`] = `Object {}`;

exports[`Writing/package json files should update package for cap projects with sap ux enabled 1`] = `
Object {
  "": Object {
    "contents": "{
  \\"name\\": \\"test_cap\\",
  \\"version\\": \\"0\\",
  \\"description\\": \\"Test file\\",
  \\"dependencies\\": {
    \\"@sap/cds\\": \\"6.8.2\\"
  },
  \\"sapux\\": [
    \\"app/test-cap-package-no-sapux\\"
  ],
  \\"scripts\\": {
    \\"watch-test-cap-package-no-sapux\\": \\"cds watch --open test-cap-package-no-sapux/webapp/index.html?sap-ui-xx-viewCache=false\\"
  }
}
",
    "state": "modified",
  },
}
`;
